### PR TITLE
hw: i386: virt: Add support for PCI hotplug through ACPI

### PR DESCRIPTION
**Pull request by sboeuf**
This PR enables the support for hotplugging PCI devices through ACPI:
- Enable PCI bus to accept hotplugged devices, by providing an
appropriate hotplug handler.
- Implement the code path generating the appropriate interrupt in
case a new PCI device is getting added to the PCI bus.
- Modify the definition of the PCI bus in the DSDT table to make
sure a PCI scan of all the available slots can be performed when
the interrupts comes in.
- Modify GED device in the DSDT table, so that an interrupt related
to PCI will trigger a new PCI scan.

Part of this work include some refactoring and the removal of unneeded
code related to the pc machine type.

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>
Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

### CODE REVIEW

**Comment by sboeuf** [comment]
@mcastelino @sameo @rbradford @liujing2 PTAL

**Review by rbradford — APPROVED** [review]

File: hw/i386/virt/acpi.c [hunk 1]
```diff
@@ -197,6 +208,14 @@ DeviceState *virt_acpi_init(qemu_irq *gsi)
 
     s = VIRT_ACPI(dev);
     s->gsi = gsi;
+    s->pci_bus = pci_bus;
```
Comment (mcastelino) [inline hw/i386/virt/acpi.c:211]: @sboeuf one small nit. Should always assume we have PCI? may make sense to do this only if  we are passed a PCI bus.

**Review by mcastelino — APPROVED** [review]
LGTM.

Left one small comment which you may want to address now or later. Basically I do not want us to assume the existence of the PCI bus by default.

File: hw/i386/virt/acpi.c (hunk 1 above)
Comment (sboeuf) [reply hw/i386/virt/acpi.c:211]: Yes you're right. I can check the value of `pci_bus`, and if it's `NULL`, I won't initialize the PCI hotplug.
Does that work for you ?

Comment (sboeuf) [reply hw/i386/virt/acpi.c:211]: fixed!

**Comment by mcastelino** [comment]
LGTM

File: hw/acpi/aml-build.c [hunk 2]
```diff
@@ -2314,9 +2315,203 @@ Aml *build_pci_host_bridge(Aml *table, AcpiPciBus *pci_host)
...
+    }
+    aml_append(parent_scope, method);
+    qobject_unref(bsel);
+}
+
 void acpi_dsdt_add_pci_bus(Aml *table, AcpiPciBus *pci_host)
 {
-    Aml *dev, *sb_scope;
+    Aml *dev;
+    Aml *sb_scope;
```
Comment (liujing2) [inline hw/acpi/aml-build.c]: @sboeuf forgive the nitpick :-p  Not a change here right?

**Review by liujing2 — COMMENTED** [review]
LGTM

**Comment by liujing2** [comment]
@sboeuf @mcastelino @sameo @rbradford 
I applied the patches onto 3.0.0, and tested virtio-net-pci hotplug. It failed.
It's not because of the patches, but I guess some changes between 2.12.0 and 3.0.0?
I opened an issue #65 
```
Nemu 2.12.0 
(qemu) device_add virtio-net-pci,id=virtio-net1,bus=pcie.0
[   15.040580] pci 0000:00:03.0: [1af4:1000] type 00 class 0x020000
[   15.040901] pci 0000:00:03.0: reg 0x10: [io  0x0000-0x001f]
[   15.040985] pci 0000:00:03.0: reg 0x14: [mem 0x00000000-0x00000fff]
[   15.041371] pci 0000:00:03.0: reg 0x20: [mem 0x00000000-0x00003fff 64bit pref]
[   15.041448] pci 0000:00:03.0: reg 0x30: [mem 0x00000000-0x0003ffff pref]
[   15.042890] pci 0000:00:03.0: BAR 6: assigned [mem 0x90040000-0x9007ffff pref]
[   15.042897] pci 0000:00:03.0: BAR 4: assigned [mem 0x800008000-0x80000bfff 64bit pref]
[   15.043049] pci 0000:00:03.0: BAR 1: assigned [mem 0x90002000-0x90002fff]
[   15.043088] pci 0000:00:03.0: BAR 0: assigned [io  0x1000-0x101f]
[   15.043186] virtio-pci 0000:00:03.0: enabling device (0000 -> 0003)
[   15.045969] virtio-pci 0000:00:03.0: can't derive routing for PCI INT A
[   15.045972] virtio-pci 0000:00:03.0: PCI INT A: no GSI
[   15.051700] virtio_net virtio2 ens3: renamed from eth0

Nemu 3.0.0 (The same image and parameters )
[   23.805642] pci 0000:00:03.0: [1af4:1000] type 00 class 0x020000
[   23.806129] pci 0000:00:03.0: reg 0x10: [io  0x0000-0x001f]
[   23.806233] pci 0000:00:03.0: reg 0x14: [mem 0x00000000-0x00000fff]
[   23.806602] pci 0000:00:03.0: reg 0x20: [mem 0x00000000-0x00003fff 64bit pref]
[   23.806716] pci 0000:00:03.0: reg 0x30: [mem 0x00000000-0x0003ffff pref]
[   23.808313] pci 0000:00:03.0: BAR 6: no space for [mem size 0x00040000 pref]
[   23.808317] pci 0000:00:03.0: BAR 6: failed to assign [mem size 0x00040000 pref]
[   23.808321] pci 0000:00:03.0: BAR 4: no space for [mem size 0x00004000 64bit pref]
[   23.808324] pci 0000:00:03.0: BAR 4: failed to assign [mem size 0x00004000 64bit pref]
[   23.808328] pci 0000:00:03.0: BAR 1: no space for [mem size 0x00001000]
[   23.808331] pci 0000:00:03.0: BAR 1: failed to assign [mem size 0x00001000]
[   23.808334] pci 0000:00:03.0: BAR 0: no space for [io  size 0x0020]
[   23.808337] pci 0000:00:03.0: BAR 0: failed to assign [io  size 0x0020]
[   23.808402] virtio-pci 0000:00:03.0: can't derive routing for PCI INT A
[   23.808405] virtio-pci 0000:00:03.0: PCI INT A: no GSI
```

**Comment by mcastelino** [comment]
@liujing2 @sboeuf will your refactored code work with Q35. Testing your refactored code with PCI hotplug using ACPI. This may be useful as it will unify the pc, q35 and virt hotplug handling paths for kata.

**Comment by sameo** [comment]
Rebased on top of the latest `topic/virt-x86` branch

**Comment by rbradford** [comment]
Are we getting close to being able to merge this?

**Comment by sameo** [comment]
With this new rebased PR, virtio PCI hotplug works for me. Once the Jenkins CI passes, I will merge that PR.

**Comment by liujing2** [comment]
The [bug](https://github.com/intel/nemu/pull/63#issuecomment-416466400) comment above is fixed by @rbradford . With a little fix patch squashed into [commit](https://github.com/intel/nemu/pull/63/commits/a73152b1ac5525237c16eda8d75043d1d125e0b5) , now virtio-pci  hotplug/unplug works well.

**Comment by sameo** [comment]
LGTM

**Comment by sameo** [comment]
LGTM

**Pull request merged**